### PR TITLE
Fixes and improves several access key (accelerator) related issues

### LINKED ISSUE #7090: Duplicate Accelerator Keys don't work for nested MenuItem Headers
**Describe the bug**

```xaml
<Menu>
    <MenuItem Header="_File">
        <MenuItem Header="_New Project"/>
        <MenuItem Header="_Open Project"/>
        <MenuItem Header="_Save Project"/>
    </MenuItem>
    <MenuItem Header="_Edit">
        <MenuItem Header="_Find"/>
        <MenuItem Header="_Replace"/>
    </MenuItem>
</Menu>
```

This is a very simple Menu that you can find in most editor-like applications. This is how I expect it to work:

Rider

Both `File` and `Find` start with `F` but they are not on the same level, `Find` is nested inside the `Edit` menu meaning you have to do `Alt+E+F` to get to `Find`. Avalonia does not work this way:

Avalonia

As you can see, if the `Edit` menu is open and I press `F`, I expect it to select the `Find` sub-menu item but instead it goes to `File`.

**Desktop (please complete the following information):**

- OS: Windows 10
- Version 0.10.10

### LINKED ISSUE #13160: Menu with duplicate accelerators always incorrectly selects the first one
## Describe the bug
A menu with duplicate accelerators, when navigated with the keyboard, will incorrectly select the first one of the duplicates instead of leaving the menu open, to allow the user to type the accelerator again to select the next item, `Enter` or `space` to select the current menu item, or `Esc` to cancel menu naviagation.

## To Reproduce
1. Create a window with the following menu:
```xml
<Menu>
  <MenuItem Header="_File">
    <MenuItem Header="E_xit" Command="{Binding FileExit}" />
  </MenuItem>
  <MenuItem Header="_View">
    <MenuItem Header="_Duplicate 1" Command="{Binding ViewDuplicate1}"/>
    <MenuItem Header="_Duplicate 2" Command="{Binding ViewDuplicate2}"/>
 </MenuItem>
</Menu>
```
2. Type `Alt+V` to bring up the `View` menu
3. Type `D` or `Alt+D`

## Result:
The menu is dismissed and the command bound to `Duplicate 1` is executed.

## Expected:
The menu to remain open, to allow a possible second `D` to select `Duplicate 2`

## Environment 

- OS: reproduced on Windows.
- Avalonia-Version: reproduced on 11.0.4

## Additional context
I've attached a zipped VS solution that exhibits the problem.
[AvaloniaMenuBug.zip](https://github.com/AvaloniaUI/Avalonia/files/12834324/AvaloniaMenuBug.zip)

### PR DESCRIPTION
**Pull request by StefanKoell**
## What does the pull request do?
This PR fixes and improves multiple issues with access keys / accelerators:
- Menus, menu items and labels with the same access keys are now handled in a better way allowing the user to "cycle" through all items which have the same access key assigned.
- Navigating within a menu using access keys didn't behave as expected - this has also been fixed.

## What is the current behavior?
- The current/incorrect behavior is that if multiple elements have the same access key, only the first element found with a matching access key in the registered elements is executed. 
- Also when the menu was opened using an accelerator key, submenu items are not selectable using accelerator keys.

This makes it hard to impossible to navigate through more complex applications using the keyboard. It also makes it harder to localize complex applications to ensure an access key is only used once. This can't really be guaranteed. 
Also if the application provides extensive customization where menus and the containing menu items are user customizable, items with the same access key may be encountered.

## What is the updated/expected behavior with this PR?
- if a focus scope contains only one item for a given accelerator, the behavior is the same as before: it will just execute the command / focus the target control.
- if a focus scope contains multiple items for a given accelerator, the first item receives the focus. Pressing the same accelerator again, sets the focus to the next item. This is done until all items have been focused. Pressing the same accelerator again when the last item is focused, sets the first item again.
- If a menu is open and the accelerators are visible, the menu items can now be executed/focused using the access key.

## How was the solution implemented (if it's not obvious)?
I think the code in the AccessKeyHandler, MenuItemAccessKeyHandler and the Label control are self-explanatory. If more details are needed, I'm happy to provide them.

I also added a couple of corner cases to the Menu page in the Control Catalog app to demonstrate the new behavior.

## Checklist

- [X] Added unit tests
- [X] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
Not that I can see.

## Fixed issues
Fixes #13160 
Fixes #7090

### CODE REVIEW

**Comment by cla-avalonia** [comment]
@hdocsek @StefanKoell,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

**0** out of **2** committers have signed the CLA.
- [ ] @hdocsek
- [ ] @StefanKoell

**Comment by StefanKoell** [comment]
@cla-avalonia agree

**Comment by hdocsek** [comment]
@cla-avalonia agree

**Review by MrJul — CHANGES_REQUESTED** [review]
Thank you for you contribution!

While a better handling of access keys is needed, I'm not sure this is the correct way to proceed. Currently access keys are independent from the focused element: anything can happen in reaction to the `AccessKeyPressedEvent`, including setting the focus to another control (`Label`), or activating itself (`TabItem`).

Now, it only works if there is a single match. When there are several, an extra match can get the focus even if it isn't the one supposed to (e.g. `Label`), and the event itself is never raised.

If the element itself isn't focusable, switching to it won't be possible. If the element was doing anything else than setting the focus in response to the event, it won't work.

Take `TabItem` for example. Before this PR, having two tab items with the same accelerator made it impossible to switch to the second one - which isn't great. After this PR, none of the items can be accessed using the accelerator, because `TabItem` doesn't have any other way to activate itself - this is probably worse.

I know it's the whole point of the PR, but this ties the focused element and the access key registration a bit too much.

**Comment by MrJul** [comment]
For reference, this is how this works in WPF:

1. The `AccessKeyPressedEvent` is raised on *all* registrations matching the pressed key.
2. Each registered element may provide a target in response, which can be different from itself.
3. Those targets are gathered.
4. If the currently focused element is one of the target, the next one is used, otherwise the first one (similar to this PR).
5. Once the final target has been chosen, an `AccessKeyEvent` (not AccessKey*Pressed*Event) is raised on that element, which is free to set the focus, activate the item, etc.

Aside from the confusing names between `AccessKeyEvent` and `AccessKeyPressedEvent`, I think it's reasonable to decouple finding the access key target (responsibility of the registered element) from responding to it (responsibility of the target element).

Source: https://github.com/dotnet/wpf/blob/13c81c7954e5195e9a9189dc9600b3cb120ed031/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/AccessKeyManager.cs

**Comment by StefanKoell** [comment]
Thanks for the feedback and the reference to the WPF implementation. I've looked at the linked AccessKeyManager implementation of WPF and reading the code, makes sense and is for sure a much better/robust implementation. Admittedly the PR we submitted did not take all those cases into consideration but for real world apps, the code from our PR works much better in more complex UIs than the current implementation. I can understand that this PR is not ready for merge because of the issues you mentioned.

The question now is, do you want me/us to look further into that and trying to replicate the behavior of WPF's approach or is this something the Avalonia core team should work on because of the implications it may have on a much lower level?

**Comment by MrJul** [comment]
> Admittedly the PR we submitted did not take all those cases into consideration but for real world apps, the code from our PR works much better in more complex UIs than the current implementation.

While I agree that the overall access key behavior is much better after your PR, I'm really concerned about the Label becoming focusable. This might break some UI and is generally unexpected. The rest of the changes is internal and should cause no issues.

> The question now is, do you want me/us to look further into that and trying to replicate the behavior of WPF's approach or is this something the Avalonia core team should work on because of the implications it may have on a much lower level?

I think replicating WPF's `AccessKeyManager` in Avalonia's `AccessKeyHandler` would be the best as it solves all problems cleanly. It should be fairly contained and we'll be happy to accept a pull request if you want to look further into this. The exact API shape should be discussed though; I'll open an issue soon for that.

**Comment by hdocsek** [comment]
@MrJul We implemented a functional draft based on the WPF behavior, so that we can have clearer picture and to provide a concrete example for discussing the exact API.

Additionally, we added a dedicated page named `Accelerator` to the catalog to demonstrate the behavior. 
It sports various duplicate access keys for elements like e.g. buttons, labels, menus, and tabs.

The basic idea is to have two events which are handled by the `AccessKeyHandler` when processing accelerator keys.
    - `AccessKeyPressedEvent`: Used to collect information about potential targets
    - `AccessKeyEvent`: Used to actually handle the accelerator key.

The `AccessKeyHandler.AccessKeyPressedEvent` decides whether/which element is used as a potential target.

In `AccessKeyHandler.OnKeyDown` these targets are collected and than it is calculated for which element the `AccessKeyHandler.AccessKeyEvent` should be invoked.

`AccessKeyHandler.AccessKeyEvent` is wired up on each `InputElement` to a virtual method named `OnAccessKey` which by default calls `Focus(NavigationMethod.Tab)`.
This method can than overridden where specific handling is needed.

Up to now we identified the following places where this needs to be done:

- `Button`
- `Label`
- `Menu`
- `MenuItem`
- `TabItem`

Please note:

- `MenuItemAccessKeyHandler` has been removed. From what we see, all scenarios can be handled with a single implementation (as it is now). 
- The `DefaultMenuInteractionHandler` already has a method for `AccessKeyPressed` which is `protected internal`. 
We didn't change the signature but the naming based on WPF implementation may cause confusion as we needed to change the event handler registration from `AccessKeyPressedEvent` to `AccessKeyEvent`.
- We noticed that the `IsKeyboardFocusWithin` of the `PopupRoot` seems to work only at the first time (for a `MenuItem`). Maybe this is a bug in `KeyboardDevice.SetIsFocusWithin(...)`. Removing `el != branch` in [KeyboardDevice](https://github.com/AvaloniaUI/Avalonia/blob/0944e043cab6d7c6443826bbfe770c309d611591/src/Avalonia.Base/Input/KeyboardDevice.cs#L100) resolves the issue, but we are not sure about side effects. Therefore we implemented our own handling in `AccessHandler.IsFocusWithinOwner(IInputRoot)` to reliably determine if the element is in scope. 
- We are sorting potential targets by hierarchy (`AccessHandler.SortByHierarchy(List<IInputElement>)`) to ensure that the targets are processed in the correct order. For example, in a TabControl the selected tab item content will be processed before the next tab item. Maybe there is a better way to handle this.
- Also note that some of the unit tests are broken now. Once we have finalized the API we will revisit and fix them. 

The design and naming is based on WPF. If you need any changes to this PR, please let us know.

**Comment by hdocsek** [comment]
@MrJul I fixed the unit tests but the PR build still fails.

I think this is due to this change in the API surface

`CP0012: Cannot remove virtual keyword from member 'Avalonia.Controls.Button.OnAccessKey(Avalonia.Interactivity.RoutedEventArgs)'.
`

When we added the virtual method `'InputElement.OnAccessKey(Avalonia.Interactivity.RoutedEventArgs)` we saw that
`Avalonia.Controls.Button` already defined it  and replaced it with an override. 

Is this an acceptable change and what needs to be done to fix above error?

**Review by MrJul — CHANGES_REQUESTED** [review]
This is looking really good to me, it's much better than the current behavior! Thanks a lot for matching the WPF implementation and adding a test page.

I've left several minor comments. The logic itself is good (aside from an issue with sorting the targets).

**Comment by MrJul** [comment]
> @MrJul I fixed the unit tests but the PR build still fails.
> 
> I think this is due to this change in the API surface
> 
> `CP0012: Cannot remove virtual keyword from member 'Avalonia.Controls.Button.OnAccessKey(Avalonia.Interactivity.RoutedEventArgs)'. `
> 
> When we added the virtual method `'InputElement.OnAccessKey(Avalonia.Interactivity.RoutedEventArgs)` we saw that `Avalonia.Controls.Button` already defined it and replaced it with an override.
> 
> Is this an acceptable change and what needs to be done to fix above error?

According to [Change rules for compatibility](https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules#members):
✔️ ALLOWED: Moving a member into a class higher in the hierarchy than the type from which it was removed

The change should be fine. As far as binary compatibility is concerned, any existing override of `Button.OnAccessKey` should have no trouble finding the new method with the same signature. (It would be worth a double check though.)

Run `nuke ValidateApiDiff` in the root folder to generate the API suppression file.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: Yes, I reverted it.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: In the WPF this held not only the target but also also the scope. I thought I keep it that way in case we want to attached additional information in the future. 
Futhermore, `GetInfoForElement` was renamed to `GetTargetForElement`.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: I removed the lock statement.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: Done.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: Done.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: I removed the lock statement.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: This was unintentional, I just tried sthg and forgot to revert.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: I changed all code parts which remove registrations to reversed for-loops.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: The private types have been moved to the bottom of the class/file.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: Changed `AccessKeyRegistration` from `record` to `class`.

File: tests/Avalonia.Controls.UnitTests/ButtonTests.cs
Comment (hdocsek) [reply tests/Avalonia.Controls.UnitTests/ButtonTests.cs]: I reverted the unneeded formatting.

File: tests/Avalonia.Controls.UnitTests/TabControlTests.cs
Comment (hdocsek) [reply tests/Avalonia.Controls.UnitTests/TabControlTests.cs]: I reverted the unneeded formatting.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: I changed the code so that it uses the `ILogical` interface instead of `InputElement`. (see [here](https://github.com/royalapplications/Avalonia/blob/36b7e77dc81c3fe76ff618528b3cfc498274ce4b/src/Avalonia.Base/Input/AccessKeyHandler.cs#L473))

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (hdocsek) [reply src/Avalonia.Base/Input/AccessKeyHandler.cs]: I simplified the sorting code for clarity, see [here](https://github.com/royalapplications/Avalonia/blob/36b7e77dc81c3fe76ff618528b3cfc498274ce4b/src/Avalonia.Base/Input/AccessKeyHandler.cs#L457), but basically it is the same procedure.
The original order of the incoming elements is the registration order, but in case of nested elements, for example in tabs, I want to run over the child elements before going to the next parent. 
The tabs on the Accelerator page in the catalog show the behavior, it feels quite intuitive.

**Comment by hdocsek** [comment]
@MrJul Thank you very much for your input. I updated the API suppression file and adressed the issues in your comments. 
Please let me know, if there is anything else to be done.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (MrJul) [inline src/Avalonia.Base/Input/AccessKeyHandler.cs:458]: Small optimization: return `targets` immediately if we have 0 or 1 items (the very common cases). There's no point in allocating the list and the queue if no sorting effectively occurs.

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (MrJul) [inline src/Avalonia.Base/Input/AccessKeyHandler.cs:500]: I've missed this previously, but please keep the event arguments `internal` for now, since the event itself is internal.

We should definitely expose the events publicly at some point to allow users to participate in access key handling, but that should be another PR with some API review. I don't want to block this PR because of that :)

File: src/Avalonia.Base/Input/AccessKeyHandler.cs
Comment (MrJul) [inline src/Avalonia.Base/Input/AccessKeyHandler.cs:537]: Same as above, `internal` for now.

**Review by MrJul — CHANGES_REQUESTED** [review]
Thank you! A couple of minor points left and it'll be good to merge.

**Review by MrJul — APPROVED** [review]

**Comment by MrJul** [comment]
Thank you, LGTM!

**Comment by hdocsek** [comment]
@MrJul Thank you very much for your help.

**Comment by TangDave** [comment]
Does it fix the problem with menu items being accessed via accelerator keys even when disabled?

**Comment by hdocsek** [comment]
@TangDave I don't know  which issue you are refering to exactly, but I did a quick check. 
Looks like this is fixed, disabled menu items cannot be triggered via accelerator keys with this PR.

**Comment by TangDave** [comment]
@hdocsek - thanks, I'll check it out when 11.3 is official

**Comment by StefanKoell** [comment]
@TangDave not sure if this change will be part of 11.3. But you can test the PR using the following build number:

> You can test this PR using the following package version. `11.3.999-cibuild0053125-alpha`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Pull request merged**
